File: app/routers/resultados.py
"""Resultados router — query and trigger indicator calculations.

Task 4.3:
- GET  /resultados?indicador_id=X&periodo_inicio=...&periodo_fin=...
       → filterable, paginated list of pre-computed results.
- POST /resultados/calcular-ahora
       → iterate all active indicators, compute their periodo dates,
         run engine/interpreter + executor, return batch summary.
"""

import uuid
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_async_session_factory
from app.engine.periodo import calcular_periodo
from app.models.indicador import Indicador, IndicadorResultado, IndicadorVersion
from app.schemas.indicador import (
    BatchCalcularNowResponse,
    ErrorCalculo,
    IndicadorResultadoEnrichedResponse,
    ResultadoListResponse,
)

router = APIRouter()


# ── Dependency ─────────────────────────────────────────────────────────


async def get_db():
    """Yield an async database session from the global factory."""
    factory = get_async_session_factory()
    async with factory() as session:
        yield session


# ── GET filterable results ─────────────────────────────────────────────


@router.get(
    "/",
    response_model=ResultadoListResponse,
    summary="Query pre-computed indicator results",
)
async def get_resultados(
    indicador_id: uuid.UUID | None = Query(
        None, description="Filter by indicator (joins through IndicadorVersion)"
    ),
    periodo_inicio: date | None = Query(
        None, description="Filter results with period_start >= this date"
    ),
    periodo_fin: date | None = Query(
        None, description="Filter results with period_end <= this date"
    ),
    page: int = Query(1, ge=1, description="Page number (1-indexed)"),
    size: int = Query(20, ge=1, le=100, description="Items per page"),
    db: AsyncSession = Depends(get_db),
):
    """Return a filterable, paginated list of computed IndicadorResultado rows.

    When indicador_id is provided, results are scoped to that indicator
    (joined through IndicadorVersion). Date filters apply to period boundaries.
    """
    # Build base query with eager-loading for relationships
    base = select(IndicadorResultado).options(
        selectinload(IndicadorResultado.indicador_version).selectinload(
            IndicadorVersion.indicador
        )
    )

    if indicador_id is not None:
        base = base.join(IndicadorVersion).where(
            IndicadorVersion.indicador_id == indicador_id
        )
    if periodo_inicio is not None:
        base = base.where(IndicadorResultado.periodo_inicio >= periodo_inicio)
    if periodo_fin is not None:
        base = base.where(IndicadorResultado.periodo_fin <= periodo_fin)

    # Count with identical filters (no selectinload needed)
    count_from = IndicadorResultado
    count_query = select(func.count()).select_from(count_from)
    if indicador_id is not None:
        count_query = count_query.join(IndicadorVersion).where(
            IndicadorVersion.indicador_id == indicador_id
        )
    if periodo_inicio is not None:
        count_query = count_query.where(
            IndicadorResultado.periodo_inicio >= periodo_inicio
        )
    if periodo_fin is not None:
        count_query = count_query.where(
            IndicadorResultado.periodo_fin <= periodo_fin
        )

    total: int = (await db.execute(count_query)).scalar() or 0
    pages: int = max(1, (total + size - 1) // size)

    # Fetch page
    result = await db.execute(
        base.order_by(IndicadorResultado.calculado_en.desc())
        .offset((page - 1) * size)
        .limit(size)
    )
    rows = result.scalars().all()

    items = [
        IndicadorResultadoEnrichedResponse(
            id=r.id,
            indicador_version_id=r.indicador_version_id,
            periodo_inicio=r.periodo_inicio,
            periodo_fin=r.periodo_fin,
            valor=float(r.valor),
            calculado_en=r.calculado_en,
            indicador_nombre=r.indicador_version.indicador.nombre,
            indicador_version_num=r.indicador_version.version,
        )
        for r in rows
    ]

    return ResultadoListResponse(
        items=items,
        total=total,
        page=page,
        size=size,
        pages=pages,
    )


# ── POST batch calculation ─────────────────────────────────────────────


@router.post(
    "/calcular-ahora",
    response_model=BatchCalcularNowResponse,
    summary="Calculate all active indicators now",
)
async def calcular_ahora(
    db: AsyncSession = Depends(get_db),
):
    """Iterate every active indicator and compute its result for its configured period.

    For each active indicator:
    1. Retrieve its latest IndicadorVersion.
    2. Parse the definicion into DefinicionIndicador (Pydantic).
    3. Calculate periodo dates from definicion.periodo.
    4. Build the parameterized MySQL query via engine/interpreter.py.
    5. Execute against OpenMRS MySQL and persist results via engine/executor.py.

    Individual indicator failures are caught and reported in `errores` —
    they never abort the batch.
    """
    # Fetch all active indicators
    result = await db.execute(
        select(Indicador).where(Indicador.activo.is_(True))
    )
    indicadores = result.scalars().all()

    calculados = 0
    errores: list[ErrorCalculo] = []
    total = len(indicadores)

    for indicador in indicadores:
        try:
            # Get latest version
            version_result = await db.execute(
                select(IndicadorVersion)
                .where(IndicadorVersion.indicador_id == indicador.id)
                .order_by(IndicadorVersion.version.desc())
                .limit(1)
            )
            latest = version_result.scalar_one_or_none()
            if latest is None:
                errores.append(
                    ErrorCalculo(
                        indicador_id=indicador.id,
                        indicador_nombre=indicador.nombre,
                        error="Sin versiones definidas",
                    )
                )
                continue

            # Parse definicion and compute period
            from app.types.definicion import DefinicionIndicador

            definicion = DefinicionIndicador.model_validate(latest.definicion)
            periodo_inicio, periodo_fin = calcular_periodo(definicion.periodo)

            # ── Resolve ordenes concept UUIDs to OpenMRS concept_ids ──
            concept_map: dict[str, int] | None = None
            ordenes = definicion.evento.ordenes if definicion.evento else None
            if ordenes:
                from app.database import get_sync_engine

                sync_engine = get_sync_engine()
                uuids = [str(f.concepto_uuid) for f in ordenes]
                with sync_engine.connect() as conn:
                    from sqlalchemy import text

                    result_rows = conn.execute(
                        text(
                            "SELECT uuid, concept_id FROM concept "
                            "WHERE uuid IN :uuids AND retired = 0"
                        ),
                        {"uuids": tuple(uuids)},
                    ).fetchall()

                resolved: dict[str, int] = {
                    row.uuid: row.concept_id for row in result_rows
                }

                missing: list[str] = []
                concept_map = {}
                for f in ordenes:
                    suuid = str(f.concepto_uuid)
                    cid = resolved.get(suuid)
                    if cid is None:
                        missing.append(suuid)
                    else:
                        concept_map[f.concepto_uuid] = cid

                if missing:
                    raise HTTPException(
                        status_code=502,
                        detail={
                            "field": "ordenes",
                            "reason": "CONCEPT_NOT_FOUND",
                            "missing": missing,
                        },
                    )

            # Build query and execute
            from app.engine.executor import execute_and_persist_async
            from app.engine.interpreter import build_query

            query_sql, params = build_query(
                definicion, periodo_inicio, periodo_fin, concept_map=concept_map
            )
            await execute_and_persist_async(
                query_sql, params, latest.id, periodo_inicio, periodo_fin
            )

            calculados += 1

        except Exception as exc:
            errores.append(
                ErrorCalculo(
                    indicador_id=indicador.id,
                    indicador_nombre=indicador.nombre,
                    error=str(exc),
                )
            )

    return BatchCalcularNowResponse(
        calculados=calculados,
        errores=errores,
        total=total,
    )

File: src/routers/resultados.ts
/**
 * Resultados router — query and trigger indicator calculations.
 *
 * - GET  /resultados?indicador_id=X&periodo_inicio=...&periodo_fin=...
 *        → filterable, paginated list of pre-computed results; by default
 *          only canonical rows are returned, include_historicos=true also
 *          returns superseded rows, version_id filters by specific version.
 * - GET  /resultados/series?indicador_id=X&anio=YYYY&granularity=mensual|...
 *        → time-series rollups from canonical monthly results.
 * - POST /resultados/calcular-ahora
 *        → iterate all active indicators, calculate for the current month,
 *          run engine/interpreter + executor with canonical semantics, return batch summary.
 * - POST /resultados/recalcular-anio
 *        → recalculate all active indicators for every month in a given year.
 */

import { Router, type Request, type Response } from "express";
import { Op } from "sequelize";
import {
  Indicador,
  IndicadorVersion,
  IndicadorResultado,
} from "../models/indicador.js";
import { parseDefinicionIndicador } from "../types/definicion.js";
import { buildQuery } from "../engine/interpreter.js";
import { executeAndPersist } from "../engine/executor.js";
import { writeCalcLog } from "../engine/calc-log.js";
import { queryMysql } from "../database/mysql.js";
import { calcularMesActual } from "../engine/periodo.js";
import { resolveConceptMap } from "../validators/openmrs.js";
import { OpenMRSUnavailableError } from "../errors.js";
import { asyncHandler } from "../middleware/async-handler.js";
import { requirePrivilege } from "../middleware/auth.js";
import { rateLimit, resetRateLimitStore } from "./resultados/rate-limit.js";
import { handleSeries } from "./resultados/series.js";
import { handleRecalcularAnio } from "./resultados/recalcular-anio.js";
import { settings } from "../config/index.js";

export const resultadosRouter: Router = Router();

// Re-export for testing
export { resetRateLimitStore };

// ── GET /resultados ────────────────────────────────────────────────────────

resultadosRouter.get(
  "/",
  asyncHandler(async (req: Request, res: Response) => {
    const indicadorId = req.query["indicador_id"] as string | undefined;
    const periodoInicioStr = req.query["periodo_inicio"] as string | undefined;
    const periodoFinStr = req.query["periodo_fin"] as string | undefined;
    const versionId = req.query["version_id"] as string | undefined;
    const includeHistoricos = req.query["include_historicos"] === "true";
    const pageRaw = req.query["page"] as string | undefined;
    const sizeRaw = req.query["size"] as string | undefined;

    let page = 1;
    let size = 20;

    if (pageRaw !== undefined && pageRaw !== "") {
      if (!/^\d+$/.test(pageRaw)) {
        res.status(422).json({
          detail: { field: "page", message: "page debe ser un número entero" },
        });
        return;
      }
      page = parseInt(pageRaw, 10);
      if (page < 1) {
        res.status(422).json({
          detail: { field: "page", message: "page debe ser mayor o igual a 1" },
        });
        return;
      }
    }

    if (sizeRaw !== undefined && sizeRaw !== "") {
      if (!/^\d+$/.test(sizeRaw)) {
        res.status(422).json({
          detail: { field: "size", message: "size debe ser un número entero" },
        });
        return;
      }
      size = parseInt(sizeRaw, 10);
      if (size < 1 || size > 100) {
        res.status(422).json({
          detail: { field: "size", message: "size debe estar entre 1 y 100" },
        });
        return;
      }
    }

    if (
      versionId !== undefined &&
      !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(versionId)
    ) {
      res.status(422).json({
        detail: { field: "version_id", message: "version_id debe ser un UUID válido" },
      });
      return;
    }

    // Build where clause. Default: canonical rows only (one per
    // indicador+month); include_historicos=true also returns superseded rows.
    const where: Record<string, unknown> = {};
    if (!includeHistoricos) {
      where["es_canonico"] = true;
    }
    if (versionId) {
      where["indicador_version_id"] = versionId;
    }
    if (periodoInicioStr) {
      where["periodo_inicio"] = { [Op.gte]: periodoInicioStr };
    }
    if (periodoFinStr) {
      where["periodo_fin"] = { [Op.lte]: periodoFinStr };
    }

    // If indicador_id is provided, filter through IndicadorVersion
    let includeOption: object | undefined;
    if (indicadorId) {
      includeOption = {
        model: IndicadorVersion,
        as: "indicador_version",
        where: { indicador_id: indicadorId },
        include: [
          {
            model: Indicador,
            as: "indicador",
          },
        ],
      };
    } else {
      includeOption = {
        model: IndicadorVersion,
        as: "indicador_version",
        include: [
          {
            model: Indicador,
            as: "indicador",
          },
        ],
      };
    }

    const { count, rows } = await IndicadorResultado.findAndCountAll({
      where,
      include: [includeOption as object].filter(Boolean),
      order: [["calculado_en", "DESC"]],
      offset: (page - 1) * size,
      limit: size,
    });

    const pages = Math.max(1, Math.ceil(count / size));

    const items = rows.map((r) => ({
      id: r.id,
      indicador_version_id: r.indicador_version_id,
      periodo_inicio: r.periodo_inicio,
      periodo_fin: r.periodo_fin,
      valor: parseFloat(String(r.valor)),
      calculado_en: r.calculado_en,
      mes_referencia: r.mes_referencia,
      es_canonico: r.es_canonico,
      indicador_nombre:
        (r as unknown as { indicador_version?: { indicador?: { nombre: string } } })
          .indicador_version?.indicador?.nombre ?? null,
      indicador_version_num:
        (r as unknown as { indicador_version?: { version: number } })
          .indicador_version?.version ?? null,
    }));

    res.json({
      items,
      total: count,
      page,
      size,
      pages,
    });
  }),
);

// ── GET /resultados/series ─────────────────────────────────────────────────

resultadosRouter.get(
  "/series",
  asyncHandler(async (req: Request, res: Response) => {
    await handleSeries(req, res);
  }),
);

// ── POST /resultados/calcular-ahora ────────────────────────────────────────

resultadosRouter.post(
  "/calcular-ahora",
  requirePrivilege(settings.openmrs_required_privilege),
  asyncHandler(async (req: Request, res: Response) => {
    const clientIp = req.ip ?? req.socket.remoteAddress ?? "unknown";
    if (!rateLimit(`calcular-ahora:${clientIp}`, 3, 60_000)) {
      res.status(429).json({
        detail: "Demasiadas solicitudes. Intentá de nuevo en un minuto.",
      });
      return;
    }

    // Preflight: fail fast with 502 when the OpenMRS MySQL database is
    // unreachable, instead of returning a 200 with every indicator errored.
    try {
      await queryMysql("SELECT 1", {});
    } catch {
      res.status(502).json({ detail: "OpenMRS no disponible" });
      return;
    }

    const indicadores = await Indicador.findAll({
      where: { activo: true },
    });

    // Always calculate for the current month (canonical monthly semantics)
    const { inicio, fin, finPersistencia, mes_referencia } = calcularMesActual();

    // Batch the latest-version lookup instead of issuing one query per
    // indicator. Ordering lets us keep the first row for each indicator.
    const indicadorIds = indicadores.map((indicador) => indicador.id);
    const versions = indicadorIds.length > 0
      ? await IndicadorVersion.findAll({
          where: { indicador_id: indicadorIds },
          order: [["indicador_id", "ASC"], ["version", "DESC"]],
        })
      : [];
    const latestVersionByIndicador = new Map<string, IndicadorVersion>();
    for (const version of versions) {
      if (!latestVersionByIndicador.has(version.indicador_id)) {
        latestVersionByIndicador.set(version.indicador_id, version);
      }
    }

    // Parse definitions and collect all order concepts before making one
    // OpenMRS lookup for the complete batch.
    const definicionByIndicador = new Map<string, ReturnType<typeof parseDefinicionIndicador>>();
    const allConceptUuids = new Set<string>();
    for (const indicador of indicadores) {
      const version = latestVersionByIndicador.get(indicador.id);
      if (!version) continue;

      const definicion = parseDefinicionIndicador(version.definicion);
      definicionByIndicador.set(indicador.id, definicion);
      for (const orden of definicion.evento?.ordenes ?? []) {
        allConceptUuids.add(orden.concepto_uuid);
      }
    }

    let globalConceptMap: Record<string, number> = {};
    if (allConceptUuids.size > 0) {
      try {
        globalConceptMap = await resolveConceptMap(Array.from(allConceptUuids));
      } catch (err) {
        if (err instanceof OpenMRSUnavailableError) {
          res.status(502).json({ detail: "OpenMRS no disponible" });
          return;
        }
        throw err;
      }
    }

    const conceptMapByIndicador = new Map<string, Record<string, number> | null>();
    const conceptErrorByIndicador = new Map<string, string>();
    for (const indicador of indicadores) {
      const definicion = definicionByIndicador.get(indicador.id);
      const ordenes = definicion?.evento?.ordenes;
      if (!ordenes || ordenes.length === 0) {
        conceptMapByIndicador.set(indicador.id, null);
        continue;
      }

      const conceptMap: Record<string, number> = {};
      const missing: string[] = [];
      for (const orden of ordenes) {
        const conceptId = globalConceptMap[orden.concepto_uuid];
        if (conceptId === undefined) missing.push(orden.concepto_uuid);
        else conceptMap[orden.concepto_uuid] = conceptId;
      }

      if (missing.length > 0) {
        conceptErrorByIndicador.set(
          indicador.id,
          `Conceptos de órdenes no encontrados: ${missing.join(", ")}`,
        );
      } else {
        conceptMapByIndicador.set(indicador.id, conceptMap);
      }
    }

    let calculados = 0;
    const errores: Array<{
      indicador_id: string;
      indicador_nombre: string;
      error: string;
    }> = [];
    const total = indicadores.length;

    for (const indicador of indicadores) {
      const latest = latestVersionByIndicador.get(indicador.id);
      try {
        if (!latest) {
          errores.push({
            indicador_id: indicador.id,
            indicador_nombre: indicador.nombre,
            error: "Sin versiones definidas",
          });
          await writeCalcLog({
            indicador_id: indicador.id,
            indicador_version_id: null,
            mes_referencia,
            error: "Sin versiones definidas",
            fuente: "calcular-ahora",
          });
          continue;
        }
        const conceptError = conceptErrorByIndicador.get(indicador.id);
        if (conceptError) {
          errores.push({
            indicador_id: indicador.id,
            indicador_nombre: indicador.nombre,
            error: conceptError,
          });
          await writeCalcLog({
            indicador_id: indicador.id,
            indicador_version_id: latest.id,
            mes_referencia,
            error: conceptError,
            fuente: "calcular-ahora",
          });
          continue;
        }

        const definicion = definicionByIndicador.get(indicador.id)!;
        const conceptMap = conceptMapByIndicador.get(indicador.id) ?? null;

        // Build query with month boundaries
        const { sql, params } = buildQuery(
          definicion,
          inicio,
          fin,
          conceptMap,
        );

        await executeAndPersist(
          sql,
          params as Record<string, unknown>,
          latest.id,
          inicio,
          finPersistencia,
          mes_referencia,
          {
            indicadorId: indicador.id,
            fuente: "calcular-ahora",
            persistirCeroSiVacio: true,
          },
        );

        calculados += 1;
      } catch (err: unknown) {
        const internalMessage =
          err instanceof Error ? err.message : "Error desconocido";
        errores.push({
          indicador_id: indicador.id,
          indicador_nombre: indicador.nombre,
          error: "Error interno durante el cálculo",
        });
        await writeCalcLog({
          indicador_id: indicador.id,
          indicador_version_id: latest?.id ?? null,
          mes_referencia,
          error: internalMessage,
          fuente: "calcular-ahora",
        });
      }
    }

    res.json({
      calculados,
      errores,
      total,
      mes_referencia: mes_referencia.toISOString().slice(0, 7),
    });
  }),
);

// ── POST /resultados/recalcular-anio ───────────────────────────────────────

resultadosRouter.post(
  "/recalcular-anio",
  requirePrivilege(settings.openmrs_required_privilege),
  asyncHandler(async (req: Request, res: Response) => {
    await handleRecalcularAnio(req, res);
  }),
);
